File: 2022_11_19_leetcode_in_csharp/LeetCode/P1365_SmallerNumbersThanCurrent.cs
// https://leetcode.com/problems/how-many-numbers-are-smaller-than-the-current-number/

namespace LeetCode;

using static System.Linq.Enumerable;

class P1365SmallerNumbersThanCurrent
{
    static int[] SmallerNumbersThanCurrent(int[] numbers)
    {
        return numbers
            .Select((n, index) => Range(0, numbers.Length).Count(j => j != index && numbers[j] < n))
            .ToArray();
    }

    public static void Run()
    {
        var inputs = new[]
        {
            new []  {8, 1, 2, 2, 3},
            new []  {6, 5, 4, 8},
            new []  {7, 7, 7, 7},
        };

        foreach (var input in inputs)
        {
            var output = SmallerNumbersThanCurrent(input);
            Console.WriteLine($" {input.JoinToString()} -> {output.JoinToString()}");
        }
    }
}

File: 2022_11_03_leetcode_in_fsharp/LeetCode/P1365_SmallerNumbersThanCurrent.fs
// url: https://leetcode.com/problems/how-many-numbers-are-smaller-than-the-current-number/
// tags: mapi, filter
// examples: [| 8; 1; 2; 2; 3 |] -> [4;0;1;1;3]

module LeetCode.P1365_SmallerNumbersThanCurrent

open Utils

let smallerNumbersThanCurrent numbers =
    numbers
    |> Seq.mapi (fun i n ->
        { 0 .. Array.length numbers - 1 } |> Seq.filter (fun j -> i <> j && numbers[j] < n) |> Seq.length)
    |> Seq.toArray


[| 8; 1; 2; 2; 3 |] |> smallerNumbersThanCurrent === [| 4; 0; 1; 1; 3 |]
[| 7; 7; 7; 7 |] |> smallerNumbersThanCurrent === [| 0; 0; 0; 0 |]
